### ISSUE TITLE
Interface for object does not enforce strict types on functions

### ISSUE THREAD

**Issue by Right2Drive** [Microsoft/TypeScript#21193]
**TypeScript Version:**  2.6.2

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
type MyType = {
  type: 'TEST_TYPE',
};

/** Generic type for action creator object */
export interface Actions<T> {
  [key: string]: (...args: any[]) => T;
}

const actions: Actions<MyType> = {
  setDrawerLocked(locked: boolean) {
    return {
      type: 'TEST_TYPE',
      test: 'SHOULD_BREAK',
    };
  },
};

```

**Expected behavior:**

I believe this should cause a compilation error, as the return type of the function does not strictly adhere to the type `MyType` (adds additional properties). This might well be by design, but if so I would greatly appreciate any guidance on how to achieve a strict return type on the `Actions` values :smile: 

**Actual behavior:**
Passes compilation without any complaints

**Related:**
